File: fixlab-web/src/app/services/reparacion.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable, inject } from '@angular/core';
import { Observable } from 'rxjs';
import { environment } from '../../environments/environment';
import {
  ReparacionAsignarTecnicoReqDTO,
  ReparacionCambiarEstadoReqDTO,
  ReparacionCreateReqDTO,
  ReparacionDiagnosticoCotizacionReqDTO,
  ReparacionEvidenciaReqDTO,
  ReparacionProductoReqDTO,
  ReparacionRespDTO,
  TallerRespDTO,
  TipoEquipoRespDTO,
  TipoTallerRespDTO,
} from '../models/reparacion.model';

@Injectable({ providedIn: 'root' })
export class ReparacionService {
  private http = inject(HttpClient);

  private readonly base = environment.apiBaseUrl
    ? `${environment.apiBaseUrl.replace(/\/$/, '')}/api/reparaciones`
    : '/api/reparaciones';

  listarTiposEquipo(): Observable<TipoEquipoRespDTO[]> {
    return this.http.get<TipoEquipoRespDTO[]>(`${this.base}/catalogo/tipos-equipo`);
  }

  listarTiposTaller(): Observable<TipoTallerRespDTO[]> {
    return this.http.get<TipoTallerRespDTO[]>(`${this.base}/catalogo/tipos-taller`);
  }

  listarTalleres(): Observable<TallerRespDTO[]> {
    return this.http.get<TallerRespDTO[]>(`${this.base}/catalogo/talleres`);
  }

  /** Cliente: propias; ADMIN/TECNICO: todas. */
  listar(): Observable<ReparacionRespDTO[]> {
    return this.http.get<ReparacionRespDTO[]>(this.base);
  }

  crear(dto: ReparacionCreateReqDTO): Observable<ReparacionRespDTO> {
    return this.http.post<ReparacionRespDTO>(this.base, dto);
  }

  obtenerPorId(id: number): Observable<ReparacionRespDTO> {
    return this.http.get<ReparacionRespDTO>(`${this.base}/${id}`);
  }

  obtenerPorNumeroTicket(numero: string): Observable<ReparacionRespDTO> {
    const n = encodeURIComponent(numero.trim());
    return this.http.get<ReparacionRespDTO>(`${this.base}/por-ticket/${n}`);
  }

  aprobarCotizacion(id: number): Observable<ReparacionRespDTO> {
    return this.http.post<ReparacionRespDTO>(`${this.base}/${id}/aprobar-cotizacion`, {});
  }

  asignarTecnico(id: number, dto: ReparacionAsignarTecnicoReqDTO): Observable<ReparacionRespDTO> {
    return this.http.patch<ReparacionRespDTO>(`${this.base}/${id}/asignar-tecnico`, dto);
  }

  registrarDiagnostico(id: number, dto: ReparacionDiagnosticoCotizacionReqDTO): Observable<ReparacionRespDTO> {
    return this.http.patch<ReparacionRespDTO>(`${this.base}/${id}/diagnostico`, dto);
  }

  cambiarEstado(id: number, dto: ReparacionCambiarEstadoReqDTO): Observable<ReparacionRespDTO> {
    return this.http.patch<ReparacionRespDTO>(`${this.base}/${id}/estado`, dto);
  }

  agregarProducto(id: number, dto: ReparacionProductoReqDTO): Observable<ReparacionRespDTO> {
    return this.http.post<ReparacionRespDTO>(`${this.base}/${id}/productos`, dto);
  }

  agregarEvidencia(id: number, dto: ReparacionEvidenciaReqDTO): Observable<ReparacionRespDTO> {
    return this.http.post<ReparacionRespDTO>(`${this.base}/${id}/evidencias`, dto);
  }
}

File: fixlab-api/src/main/java/com/software/fixlab/service/interfaces/ReparacionService.java
package com.software.fixlab.service.interfaces;

import com.software.fixlab.dto.req.*;
import com.software.fixlab.dto.resp.*;
import com.software.fixlab.entity.RolUsuario;

import java.util.List;

public interface ReparacionService {

    List<TipoEquipoRespDTO> listarTiposEquipo();

    List<TipoTallerRespDTO> listarTiposTaller();

    List<TallerRespDTO> listarTalleres();

    ReparacionRespDTO crear(ReparacionCreateReqDTO dto, String emailUsuario, RolUsuario rol);

    List<ReparacionRespDTO> listar(String emailUsuario, RolUsuario rol);

    ReparacionRespDTO obtenerPorId(Integer id, String emailUsuario, RolUsuario rol);

    ReparacionRespDTO obtenerPorNumeroTicket(String numeroTicket, String emailUsuario, RolUsuario rol);

    ReparacionRespDTO asignarTecnico(Integer id, ReparacionAsignarTecnicoReqDTO dto, String emailUsuario, RolUsuario rol);

    ReparacionRespDTO registrarDiagnosticoCotizacion(Integer id, ReparacionDiagnosticoCotizacionReqDTO dto,
                                                     String emailUsuario, RolUsuario rol);

    ReparacionRespDTO cambiarEstado(Integer id, ReparacionCambiarEstadoReqDTO dto, String emailUsuario, RolUsuario rol);

    ReparacionRespDTO agregarProducto(Integer id, ReparacionProductoReqDTO dto, String emailUsuario, RolUsuario rol);

    ReparacionRespDTO agregarEvidencia(Integer id, ReparacionEvidenciaReqDTO dto, String emailUsuario, RolUsuario rol);

    ReparacionRespDTO aprobarCotizacion(Integer id, String emailUsuario, RolUsuario rol);
}
